test/sqllogictest/relation-cse.slt

```diff
@@ -264,6 +264,7 @@ EXPLAIN SELECT * FROM t1 WHERE f1 = (SELECT f1 FROM t1) AND EXISTS (SELECT f1 FR
 %6 =
 | Join %1 %4 %5 (= #0 #2)
 | | implementation = Differential %4 %5.() %1.(#0)
+| Filter !(isnull(#0))
```

Comment:
This Filter is not strictly necessary: #0 is being compared to something else during the Join, so it can not be null post-join?

Reply:
The join equivalence uses Rust-equality semantics, not the SQL comparison one, although I thinks it is guaranteed by #2 being non-null. If so, this transform should have discovered that.

Reply:
As shown by the test I just pushed, the transform works as expected. This filter is not removed because it happens to be before the join when this transform is run, but then it is lifted later on top of the join by `JoinImplementation` transform.

I'll add a new cleanup pass after `JoinImplementation` happens, which should fix this.

Reply:
Hmm, the hypothesis above is wrong, since predicate propagation already happens after `JoinImplementation`:

```
                    Box::new(crate::join_implementation::JoinImplementation::default()),
                    Box::new(crate::predicate_propagation::PredicateKnowledge),
```

I'll keep digging. Thanks for spotting this.

Reply:
```
%0 =                                                                                                                                                                                                                                                                                                                                                                                                                             
| Get ? (u201)                                                                                                                                                                                                                                                                                                                                                                                                                   
| | types = (Int32?, Int32?)                                                                                                                                                                                                                                                                                                                                                                                                     
| | keys = ()                                                                                                                                                                                                                                                                                                                                                                                                                    
| Filter !(isnull(#0))                                                                                                                                                                                                                                                                                                                                                                                                           
| | types = (Int32, Int32?)                                                                                                                                                                                                                                                                                                                                                                                                      
| | keys = ()                                                                                                                                                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                                                                                                 
%1 =                                                                                                                                                                                                                                                                                                                                                                                                                             
| Get ? (u201)                                                                                                                                                                                                                                                                                                                                                                                                                   
| | types = (Int32?, Int32?)                                                                                                                                                                                                                                                                                                                                                                                                     
| | keys = ()                                                                                                                                                                                                                                                                                                                                                                                                                    
| Filter !(isnull(#0))                                                                                                                                                                                                                                                                                                                                                                                                           
| | types = (Int32, Int32?)                                                                                                                                                                                                                                                                                                                                                                                                      
| | keys = ()                                                                                                                                                                                                                                                                                                                                                                                                                    
| Project (#0)                                                                                                                                                                                                                                                                                                                                                                                                                   
| | types = (Int32)                                                                                                                                                                                                                                                                                                                                                                                                              
| | keys = ()                                                                                                                                                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                                                                                                 
%2 =                                                                                                                                                                                                                                                                                                                                                                                                                             
| Get ? (u201)                                                                                                                                                                                                                                                                                                                                                                                                                   
| | types = (Int32?, Int32?)                                                                                                                                                                                                                                                                                                                                                                                                     
| | keys = ()                                                                                                                                                                                                                                                                                                                                                                                                                    
| Project ()                                                                                                                                                                                                                                                                                                                                                                                                                     
| | types = ()                                                                                                                                                                                                                                                                                                                                                                                                                   
| | keys = ()                                                                                                                                                                                                                                                                                                                                                                                                                    
| Reduce group=()                                                                                                                                                                                                                                                                                                                                                                                                                
| | agg count(true)                                                                                                                                                                                                                                                                                                                                                                                                              
| | types = (Int64)                                                                                                                                                                                                                                                                                                                                                                                                              
| | keys = (())                                                                                                                                                                                                                                                                                                                                                                                                                  
| Filter (err: more than one record produced in subquery), (#0 > 1)                                                                                                                                                                                                                                                                                                                                                              
| | types = (Int64)                                                                                                                                                                                                                                                                                                                                                                                                              
| | keys = (())                                                                                                                                                                                                                                                                                                                                                                                                                  
| Project ()                                                                                                                                                                                                                                                                                                                                                                                                                     
| | types = ()                                                                                                                                                                                                                                                                                                                                                                                                                   
| | keys = (())                                                                                                                                                                                                                                                                                                                                                                                                                  
| Map (err: more than one record produced in subquery)                                                                                                                                                                                                                                                                                                                                                                           
| | types = (Int32)                                                                                                                                                                                                                                                                                                                                                                                                              
| | keys = (())                                                                                                                                                                                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                                                                                                                                 
%3 =                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
| Union %1 %2                                                                                                                                                                                                                                                                                                                                                                                                                    
| | types = (Int32)                                                                                                                                                                                                                                                                                                                                                                                                              
| | keys = ()                                                                                                                                                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                                                                                                 
%4 =                                                                                                                                                                                                                                                                                                                                                                                                                             
| Get ? (u201)                                                                                                                                                                                                                                                                                                                                                                                                                   
| | types = (Int32?, Int32?)                                                                            
| | keys = ()                                                                                           
| Project ()                                                                                                                                                                                                    
| | types = ()                                                                                                                                                                                                  
| | keys = ()                                                                                                                                                                                                   
| Distinct group=()                                                                                                                                                                                                                                                                                                                                                                                                              
| | types = ()                                                                                          
| | keys = (())                                                                                         
                                                                                                                                                                                                                                                                            
%5 =                                                                                                    
| Join %0 %3 %4 (= #0 #2)                                                                                                                                                                                                                                                   
| | implementation = Unimplemented                                                                      
| | types = (Int32, Int32?, Int32)                                                                                                                                                                              
| | keys = ()                                                                             
| Project (#0, #1)                                                                                                                                                                                              
| | types = (Int32, Int32?)                                                                             
| | keys = ()                                                                                           
                                                                                                                                                            
------------------------------------------------------------------------------------------------------------------------------------------------                                                                                                                                                                        
JoinImplementation { recursion_guard: RecursionGuard { depth: RefCell { value: 0 }, limit: 2048 } }                                                                                                                                                                                                                     
------------------------------------------------------------------------------------------------------------------------------------------------                                                                                                                                                                        
%0 =                                                                                                                                                                                                                                                                                                                    
| Get ? (u201)                                                                                                                                              
| | types = (Int32?, Int32?)                                                                                                                                
| | keys = ()                                                                                                                                               
| ArrangeBy (#0)                                                                                                                                            
| | types = (Int32?, Int32?)                                                                                                                                
| | keys = ()                                                                                                                                               

%1 =                                                                                                                                                        
| Get ? (u201)                                                                                                                                              
| | types = (Int32?, Int32?)                                                                                                                                
| | keys = ()                                                                                                                                               
| Filter !(isnull(#0))                                                                                                                                      
| | types = (Int32, Int32?)                                                                                                                                 
| | keys = ()                                                                                                                                               
| Project (#0)                                                                                                                                              
| | types = (Int32)                                                                                                                                         
| | keys = ()                                                                                                                                               

%2 =                                                                                                                                                        
| Get ? (u201)                                                                                                                                              
| | types = (Int32?, Int32?)                                                                                                                                
| | keys = ()                                                                                                                                               
| Project ()                                                                                                                                                
| | types = ()                                                                                                                                              
| | keys = ()                                                                                                                                               
| Reduce group=()                                                                                                                                           
| | agg count(true)                                                                                                                                         
| | types = (Int64)                                                                                                                                         
| | keys = (())                                                                                                                                             
| Filter (err: more than one record produced in subquery), (#0 > 1)                                                                                         
| | types = (Int64)                                                                                                                                         
| | keys = (())                                                                                                                                             
| Project ()                                                                                                                                                
| | types = ()                                                                                                                                              
| | keys = (())                                                                                                                                             
| Map (err: more than one record produced in subquery)                                                                                                      
| | types = (Int32)                                                                                                                                         
| | keys = (())                                                                                                                                             

%3 =                                                                                                                                                        
| Union %1 %2                                                                                                                                               
| | types = (Int32)                                                                                                                                         
| | keys = ()                                                                                                                                               

%4 =                                                                                                                                                        
| Get ? (u201)                                                                                                                                              
| | types = (Int32?, Int32?)                                                                                                                                
| | keys = ()                                                                                                                                               
| Project ()                                                                                                                                                
| | types = ()                                                                                                                                              
| | keys = ()                                                                                                                                               
| Distinct group=()                                                                                                                                         
| | types = ()                                                                                                                                              
| | keys = (())                                                                                                                                             
| ArrangeBy ()                                                                                                                                              
| | types = ()                                                                                                                                              
| | keys = (())                                                                                                                                             

%5 =                                                                                                                                                        
| Join %0 %3 %4 (= #0 #2)                                                                                                                                   
| | implementation = Differential %3 %4.() %0.(#0)                                                                                                          
| | types = (Int32?, Int32?, Int32)                                                                                                                         
| | keys = ()                                                                                                                                               
| Map                                                                                                                                                       
| | types = (Int32?, Int32?, Int32)                                                                                                                         
| | keys = ()                                                                                                                                               
| Filter !(isnull(#0))                                                                                                                                      
| | types = (Int32, Int32?, Int32)                                                                                                                          
| | keys = ()                                                                                                                                               
| Project (#0..#2)                                                                                                                                          
| | types = (Int32, Int32?, Int32)                                                                                                                          
| | keys = ()                                                                                                                                               
| Project (#0, #1)                                                                                                                                          
| | types = (Int32, Int32?)                                                                                                                                 
| | keys = ()                                                                                                                                               

```

Oh, the hypothesis was right. The problem is that the `!isnull(#2)` predicate is not lifted from all branches of the union in `%3`.

I'll use the type information to propagate non-null constraints.

Reply:
Just pushed a fix for this. Thanks for spotting it!